src/async_fuse/memfs/cache/write_back_task.rs

```diff
@@ -0,0 +1,239 @@
+            });
+        }
+    }
+
+    sender
+        .send(())
+        .unwrap_or_else(|()| warn!("The receiver of pending block is closed unexpectedly."));
+}
+
+/// Evict blocks from the cache, if it hits the soft limit.
+async fn do_evict<P, S>(limit: SoftLimit, cache: Arc<MemoryCache<P, S>>)
+where
+    P: EvictPolicy<BlockCoordinate> + Send + Sync,
+    S: Storage + Send + Sync,
+{
+    // `a` is a numerator while `b` is a denominator, which are to represent the soft limit.
+    // If `size > capacity * a / b`, the soft limit is considered to be hit.
+    // We use the equivalent inequation `size * b / a > capacity` to avoid overflowing.
+    let SoftLimit(a, b) = limit;
+    let b = b.get();
+
+    loop {
+        let capacity = cache.policy().capacity();
+        let size = cache.policy().size();
+
+        // Soft limitation is hit, do evict.
+        if size.overflow_mul(b).overflow_div(a) > capacity {
+            cache.evict().await;
+        } else {
+            break;
+        }
+    }
+}
+
+/// A function to run a write back task.
+///
+/// This task will smoothly write blocks back to the backend
+/// when the cache runs in write-back policy.
+///
+/// And this task will evict blocks to the backend if the number of blocks hits the soft limit,
+/// whether the cache runs in write-back policy or not.
+pub(super) async fn run_write_back_task<P, S>(
+    limit: SoftLimit,
+    interval: Duration,
+    storage: Weak<MemoryCache<P, S>>,
+    mut command_receiver: mpsc::Receiver<Command>,
+    command_queue_limit: usize,
+) where
+    P: EvictPolicy<BlockCoordinate> + Send + Sync,
+    S: Storage + Send + Sync,
+{
+    let period = interval;
+    let mut interval = tokio::time::interval(period);
+
+    let mut lru_queue = LinkedHashSet::new();
+    let mut pending_blocks: HashMap<INum, HashMap<BlockId, (Block, oneshot::Sender<()>)>> =
+        HashMap::new();
+
+    loop {
+        // Tune the interval according the payload.
+        let pending_length = lru_queue.len();
+        let new_period = calculate_period(pending_length, command_queue_limit, period);
+        if interval.period() != new_period {
+            interval = tokio::time::interval(new_period);
+        }
+
+        select! {
+            command = command_receiver.recv() => {
+                match command {
+                    Some(Command::Store { coord: BlockCoordinate(ino, block_id), block, sender }) => {
+                        // Add the block to `lru_queue` and `pending_blocks`
+                        lru_queue.insert(BlockCoordinate(ino, block_id));
+                        pending_blocks.entry(ino).or_default().insert(block_id, (block, sender));
+                    }
+                    Some(Command::Flush(ino)) => {
+                        if let Some(cache) = storage.upgrade() {
+                            flush_file(&mut lru_queue, &mut pending_blocks, cache, ino).await;
+                        } else {
+                            return;
```

Comment:
It should panic here.

Reply:
A `Flush` command was sent, but the `MemoryCache` is dropped before the task can handle the command, then the upgrade will fail. Do you think this should be an error? Or this is unreachable?